src/triage/component/catwalk/storage.py

```diff
@@ -124,31 +137,48 @@ class MatrixStore(object):
     def __init__(self, matrix_path=None, metadata_path=None):
         self.matrix_path = matrix_path
         self.metadata_path = metadata_path
-        self._matrix = None
-        self._metadata = None
-        self._head_of_matrix = None
+        self.matrix = None
+        self.metadata = None
+        self.head_of_matrix = None
 
     @property
     def matrix(self):
-        if self._matrix is None:
-            self._load()
-        return self._matrix
+        if self.__matrix is None:
+            self.__matrix = self._load()
+        return self.__matrix
+
+    @matrix.setter
+    def matrix(self, matrix):
+        self.__matrix = matrix
```

Comment:
I think we'd do better to define these via `cachedproperty` from [Dickens](https://github.com/dssg/dickens) 😉 

Triage is already making use of these, for example in `triage.experiments.base`.

For example:

    from descriptors import cachedproperty

    …

    @cachedproperty
    def matrix(self):
        return self._load()

…And, that's it. You then no longer need to define any `setter` nor initialize any pseudo-hidden attribute in `__init__`. It works exactly like any read-only `property`, except that repeated access will hit a _data_ attribute named `matrix` cached in the object dictionary; (and, because `cachedproperty` does not define a setter, the data can always be set explicitly, as you've had to enable here by defining your `property` setter).

Reply:
Will cached property still let us perform the pickle customisation in this class?

Reply:
As in, overriding `__getstate__`?

`cachedproperty` doesn't fundamentally change that you're lazily caching these data on the object, but that you don't want to preserve the cache in the pickle. Mostly it tidies up all the code and references to keep track of. The only important difference, perhaps, is that it doesn't lean on the data key having an initial value of `None`; rather, `None` is an allowable and cacheable result.

I'd commented on what we're doing in `__getstate__`, and gave an example specifically catered to `cachedproperty`. Let me know that that makes sense.